File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/ConstructorKey.java
package pers.cxd.corelibrary.util.reflection;

import java.util.Arrays;
import java.util.Objects;

public class ConstructorKey {

    Class<?> clazz;
    Class<?>[] parameterTypes;
    private boolean inUse;
    private ConstructorKey next;

    private ConstructorKey(Class<?> clazz, Class<?>[] parameterTypes) {
        this.clazz = clazz;
        this.parameterTypes = parameterTypes;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ConstructorKey that = (ConstructorKey) o;
        return Objects.equals(clazz, that.clazz) &&
                Arrays.equals(parameterTypes, that.parameterTypes);
    }

    @Override
    public int hashCode() {
        int result = Objects.hash(clazz);
        result = 31 * result + Arrays.hashCode(parameterTypes);
        return result;
    }

    public void markInUse() {
        inUse = true;
    }

    private static ConstructorKey sPool;
    private static int sPoolSize;
    private static final int MAX_POOL_SIZE = 10;

    public static ConstructorKey obtain(Class<?> clazz, Class<?>[] parameterTypes){
        synchronized (ConstructorKey.class) {
            if (sPool != null) {
                ConstructorKey m = sPool;
                sPool = m.next;
                m.next = null;
                sPoolSize--;
                m.clazz = clazz;
                m.parameterTypes = parameterTypes;
                return m;
            }
        }
        return new ConstructorKey(clazz, parameterTypes);
    }

    public void recycle(){
        if (inUse){
            throw new IllegalStateException(this + " is in use, can't recycle");
        }
        clazz = null;
        parameterTypes = null;
        synchronized (ConstructorKey.class) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool;
                sPool = this;
                sPoolSize++;
            }
        }
    }

}

File: corelibrary/src/main/java/pers/cxd/corelibrary/util/reflection/ConstructorKey.kt
package pers.cxd.corelibrary.util.reflection

import java.util.*

class ConstructorKey private constructor(
    var clazz: Class<*>?,
    var parameterTypes: Array<out Class<*>?>
) {
    private var inUse = false
    private var next: ConstructorKey? = null
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as ConstructorKey
        return clazz == that.clazz &&
                parameterTypes.contentEquals(that.parameterTypes)
    }

    override fun hashCode(): Int {
        var result = Objects.hash(clazz)
        result = 31 * result + parameterTypes.contentHashCode()
        return result
    }

    fun markInUse() {
        inUse = true
    }

    fun recycle() {
        check(!inUse) { "$this is in use, can't recycle" }
        clazz = null
        parameterTypes = emptyArray()
        synchronized(ConstructorKey::class.java) {
            if (sPoolSize < MAX_POOL_SIZE) {
                next = sPool
                sPool = this
                sPoolSize++
            }
        }
    }

    companion object {
        private const val MAX_POOL_SIZE = 10
        private var sPool: ConstructorKey? = null
        private var sPoolSize = 0
        fun obtain(clazz: Class<*>?, vararg parameterTypes: Class<*>?): ConstructorKey {
            synchronized(ConstructorKey::class.java) {
                if (sPool != null) {
                    val m = sPool
                    sPool = m!!.next
                    m.next = null
                    sPoolSize--
                    m.clazz = clazz
                    m.parameterTypes = parameterTypes
                    return m
                }
            }
            return ConstructorKey(clazz, parameterTypes)
        }
    }
}